lib/cn/ltok.c

```diff
@@ -0,0 +1,203 @@
+bool
+allow_shared(struct ltok *ltok, uint8_t *counter)
+{
+    if (ltok->ltok_active_join || ltok->ltok_active_split || ltok->ltok_reserved)
+        return false;
+
+    if (*counter == UINT8_MAX)
+        return false;
+
+    *counter += 1;
+    return true;
+}
+
+static inline
+bool
+allow_exclusive(struct ltok *ltok, bool *granted)
+{
+    if (is_idle(ltok) && !ltok->ltok_reserved) {
+        *granted = true;
+        return true;
+    }
+
+    return false;
+}
+
+bool
+ltok_get(struct ltok *ltok, enum ltok_token token)
+{
+    bool granted = false;
+
+    spin_lock(&ltok->ltok_lock);
+
+    switch (token) {
+
+    case ltok_spill:
+        granted = allow_shared(ltok, &ltok->ltok_active_spills);
+        break;
+
+    case ltok_kcomp:
+        granted = allow_shared(ltok, &ltok->ltok_active_kcomps);
+        break;
+
+    case ltok_kvcomp:
+        granted = allow_shared(ltok, &ltok->ltok_active_kvcomps);
+        break;
+
+    case ltok_split:
+        granted = allow_exclusive(ltok, &ltok->ltok_active_split);
+        break;
+
+    case ltok_join:
+        granted = allow_exclusive(ltok, &ltok->ltok_active_join);
+        break;
+    }
+
+    spin_unlock(&ltok->ltok_lock);
+
+    return granted;
+}
+
+void
+ltok_put(struct ltok *ltok, enum ltok_token token)
+{
+    spin_lock(&ltok->ltok_lock);
+
+    switch (token) {
+
+    case ltok_spill:
+        assert(ltok->ltok_active_spills > 0);
+        ltok->ltok_active_spills -= 1;
+        break;
+
+    case ltok_kcomp:
+        assert(ltok->ltok_active_kcomps > 0);
+        ltok->ltok_active_kcomps -= 1;
+        break;
+
+    case ltok_kvcomp:
+        assert(ltok->ltok_active_kvcomps > 0);
+        ltok->ltok_active_kvcomps -= 1;
+        break;
+
+    case ltok_split:
+        assert(ltok->ltok_active_split);
+        ltok->ltok_active_split = false;
+        break;
+
+    case ltok_join:
+        assert(ltok->ltok_active_join);
+        ltok->ltok_active_join = false;
+        break;
+    }
+
+    /* Notify reservation if exclusive token is available */
+    if (ltok->ltok_reserved && ltok->ltok_signal && is_idle(ltok)) {
+        atomic_set(ltok->ltok_signal, 1);
+    }
+
+    spin_unlock(&ltok->ltok_lock);
+}
+
+bool
+ltok_reserve(struct ltok *ltok, atomic_int *signal)
+{
+    bool granted = false;
+
+    spin_lock(&ltok->ltok_lock);
+
+    if (!ltok->ltok_reserved) {
+        granted = true;
+        ltok->ltok_reserved = true;
+        ltok->ltok_signal = signal;
+        if (signal && is_idle(ltok))
+            atomic_set(signal, 1);
+    }
+
+    spin_unlock(&ltok->ltok_lock);
+
+    return granted;
+}
+
+void
+ltok_cancel_reservation(struct ltok *ltok)
+{
+    spin_lock(&ltok->ltok_lock);
+
+    ltok->ltok_reserved = false;
+    ltok->ltok_signal = NULL;
+
+    spin_unlock(&ltok->ltok_lock);
+}
+
+bool
+ltok_get_reserved(struct ltok *ltok, enum ltok_token token)
+{
+    bool granted = false;
+
+    spin_lock(&ltok->ltok_lock);
+
+    if (is_idle(ltok) && ltok->ltok_reserved) {
+
```

Comment:
I usually leave space around calls to lock/unlock to make them obvious unless it's a tiny piece of code in the lock.

Reply:
Then we should standardize on braces on the next line. Too many coding styles.

Reply:
Thing's have seriously gone awry if we cannot give a little leeway to the implementor as to whether or not to leave a blank line...

Reply:
You could say that about anything.

"Things have seriously gone awry if we cannot give a little leeway to the implementor as to whether or not to leave case labels at the same indentation as the switch"

Reply:
Not true.  A blank line gives the reader a little break from non-stop lines of code (best used to group short stanzas of related code), whereas indenting case labels implies a block scope that simply doesn't exist...

Agreed, code should read like a book, books have lots of white space to offset ideas and change in direction and to give the reader a break.  The thing is, code is much more dense than your typical book, and hence deserves more attention to judiciously placed white space...

That said, I only rarely agree with the blank lines that Alex oftens leaves after the start of a new block, just depends on how dense or complicated the code is...  I.e., it's a judgement call which no strict style guidelines nor automaton should be allowed to second guess..

Reply:
> Agreed, code should read like a book, books have lots of white space to offset ideas and change in direction and to give the reader a break.

The white space in the book is ideally consistent, unlike our use of whitespace in this codebase.

Reply:
This debate will never end.

And as a general rule, I think it's best practice to refrain from reformatting code just because it doesn't match your preferred style.

If you guys really want a formatter, then we can go there again, but I'd be inclined to make a rule that there no format exceptions, because otherwise, each person adds their own exceptions and we're right back where we started.